src/middlewares/createRouterMiddleware.ts

```diff
@@ -5,45 +5,59 @@ import {
   StateMapping,
   UiState,
   InternalMiddleware,
-  RouteState,
+  CreateURL,
 } from '../types';
 import { isEqual } from '../lib/utils';
 
-export type RouterProps = {
-  router?: Router;
-  stateMapping?: StateMapping;
+export type RouterProps<
+  TUiState extends UiState = UiState,
+  TRouteState = TUiState
+> = {
+  router?: Router<TRouteState>;
+  stateMapping?: StateMapping<TUiState, TRouteState>;
 };
 
-export type RoutingManager = (props?: RouterProps) => InternalMiddleware;
-
-export const createRouterMiddleware: RoutingManager = (props = {}) => {
+export const createRouterMiddleware = <
+  TUiState extends UiState = UiState,
+  TRouteState = TUiState
+>(
+  props: RouterProps<TUiState, TRouteState> = {}
+): InternalMiddleware<TUiState> => {
   const {
-    router = historyRouter(),
-    stateMapping = simpleStateMapping(),
+    router = historyRouter<TRouteState>(),
+    // technically this is wrong, as without stateMapping parameter given, the routeState *must* be UiState
```

Comment:
Can you explain a bit more?

Reply:
simpleStateMapping is a `StateMapping<UiState, UiState>`, and `createRouterMiddleware` can be called with a different generic than UiState, so the types don't really match. The only thing I can think of otherwise is only conditionally allowing stateMapping to be conditional, but I couldn't find how

Reply:
I tried things with branding, but still could not get anything working as soon as simpleStatemapping is actually used (as it uses UiState (`const stateMapping: StateMapping<UiState, TRouteState> | StateMapping<TRouteState, TRouteState>` becomes a union, and even though in that case both types _are_ the same, typescript doesn't seem to be able to know that)

<details>
<summary>I tried</summary>

```ts

export type RouterProps<
  TUiState extends UiState = UiState,
  TRouteState = UiState & { ___default: true }
> = TUiState extends { ___default: true }
  ? {
      router?: Router<TRouteState>;
      stateMapping?: never;
    }
  : {
      router?: Router<TRouteState>;
      stateMapping: StateMapping<TUiState, TRouteState>;
    };

export const createRouterMiddleware = <
  TUiState extends UiState = UiState,
  TRouteState = UiState & { ___default: true }
>(
  props: RouterProps<TUiState, TRouteState> = {} as RouterProps<
    TUiState,
    TRouteState
  >
): InternalMiddleware<TUiState> => {
  const {
    router = historyRouter<TRouteState>(),
    stateMapping = simpleStateMapping<TRouteState>(),
  } = props;
```

and variations of it

</details>